pkg/pillar/cmd/volumemgr/handleverifier.go

```diff
@@ -88,63 +99,120 @@ func MaybeRemoveVerifyImageConfig(ctx *volumemgrContext, objType, imageSha strin
 	}
 	if m.RefCount == 0 {
 		log.Fatalf("MaybeRemoveVerifyImageConfig: Attempting to reduce "+
-			"0 RefCount. Image Details - Name: %s, ImageID: %s, "+
-			"ImageSha256:%s, IsContainer: %t",
-			m.Name, m.ImageID, m.ImageSha256, m.IsContainer)
+			"0 RefCount. Image Details - Name: %s, "+
+			"ImageSha256:%s", m.Name, m.ImageSha256)
 	}
 	m.RefCount -= 1
 	log.Infof("MaybeRemoveVerifyImageConfig: RefCount to %d for %s",
 		m.RefCount, imageSha)
-	if m.RefCount == 0 {
-		unpublishVerifyImageConfig(ctx, objType, m.Key())
-	} else {
-		publishVerifyImageConfig(ctx, objType, m)
-	}
+	publishVerifyImageConfig(ctx, objType, m)
 	log.Infof("MaybeRemoveVerifyImageConfig done for %s", imageSha)
 }
 
+// MaybeDeleteVerifyImageConfig checks the refcount and if it is zero it
+// initiates the delete handshake with the verifier. That handshake occurs
+// after a MaybeRemoveVerifyImageConfig has dropped the refcount to zero
+// thus after:
+// 1. volumemgr publishes the VIC with RefCount=0
+// 2. verifier publishes the VIS with RefCount=0
+// At that point in time MaybeAddVerifyImageConfig can be called to increment
+// the refcount, but if MaybeDeleteVerifyImageConfig is called we proceed with
+// 3. volumemgr publishes the VIC with Expired=true (RefCount=0)
+// 4. verifier publishes the VIS with Expired=true (RefCount=0) in response
+// 5. handleVerifyImageStatusModify will check if a new VIC has been created
+// in volumemgr post #4. If VIC has been recreated, it will do nothing.
+// Otherwise it unpublishes the VIC. (A recreated VIC has Expired=false)
+//
+// Note that a VIC with Expired=true is effectively ignored; a replacement
+// VIC is created should MaybeAddVerifyImage be called after #3.
+// Also, any code doing a lookupVerifyImageStatus should ignore an Expired result
+//
+// 6. upon seeing the unpublish of the VIC the verifier deletes file and unpublishes the VIS
+// at this point in time verifier might see a new VIC from volumemgr if it recreated it
+// 7. handleVerifyImageStatusDelete will the unpublish the VIC if it has Expired set (if it was recreated it will not have Expired set)
+func MaybeDeleteVerifyImageConfig(ctx *volumemgrContext, objType, imageSha string) {
```

Comment:
I couldn't find where this method is being called. Is that part of WIP?

Reply:
Yes.